File: frontend/src/pages/short-drama/utils/overviewExportMarkdown.ts
import type { PipelineSummaryDto, SegmentPlanItemDto, StoryBlueprintDto } from '@/types/shortDramaApi';
import { segmentScriptDtoToStepSegmentViewModel } from './stepFourAdapters';

export function safeProjectExportBaseName(projectName: string | undefined | null): string {
  const s = (projectName ?? '')
    .replace(/[/\\?%*:|"<>]/g, '_')
    .replace(/\s+/g, ' ')
    .trim();
  return s || 'project';
}

function segmentPlanForIndex(
  blueprint: StoryBlueprintDto | null | undefined,
  segmentId: string,
  index: number,
): SegmentPlanItemDto | undefined {
  const plan = blueprint?.segment_plan;
  if (!Array.isArray(plan) || plan.length === 0) return undefined;
  const byId = plan.find((p) => (p.segment_id ?? '').trim() === segmentId.trim());
  if (byId) return byId;
  return plan[index];
}

function formatProductExposure(plan: SegmentPlanItemDto | undefined): string | undefined {
  if (!plan) return undefined;
  const m = toMarkdownText(plan.product_exposure_mode);
  if (m) return m;
  const s = toMarkdownText(plan.summary);
  if (s) return s;
  return undefined;
}

const toMarkdownText = (value: string | string[] | undefined | null): string => {
  if (Array.isArray(value)) return value.filter(Boolean).join('、');
  return typeof value === 'string' ? value.trim() : '';
};

/** 无数据时返回 null，由调用方提示「数据不完整」 */
export function buildOverviewScriptMarkdown(pipeline: PipelineSummaryDto | null): string | null {
  if (!pipeline?.project) return null;
  const rows = pipeline.segment_scripts;
  if (!Array.isArray(rows) || rows.length === 0) return null;

  const name = safeProjectExportBaseName(pipeline.project.project_name);
  const p = pipeline.project;
  const bp = pipeline.story_blueprint?.blueprint ?? null;

  const lines: string[] = [];
  lines.push(`# ${name} — 脚本`);
  lines.push('');
  lines.push('## 项目信息');
  lines.push(`- **项目名**：${name}`);
  lines.push(`- **格式**：${p.format?.trim() || '—'}`);
  lines.push(`- **风格**：${toMarkdownText(p.style) || '—'}`);
  lines.push(`- **视听风格**：${p.visual_style?.trim() || '—'}`);
  lines.push(`- **比例**：${p.aspect_ratio?.trim() || '—'}`);
  lines.push(`- **时长**：${p.duration?.trim() || '—'}`);
  lines.push('');

  const sorted = [...rows].sort((a, b) => {
    const na = parseInt(String(a.segment_id).replace(/\D/g, ''), 10) || 0;
    const nb = parseInt(String(b.segment_id).replace(/\D/g, ''), 10) || 0;
    return na - nb;
  });

  sorted.forEach((row, idx) => {
    const vm = segmentScriptDtoToStepSegmentViewModel(row, idx);
    const plan = segmentPlanForIndex(bp, row.segment_id, idx);
    const label = `S${idx + 1}`;
    lines.push(`## ${label} — ${vm.name}`);
    lines.push(`- **标题**：${vm.name}`);
    lines.push(`- **时长**：${vm.duration}`);
    lines.push(`- **情绪 / 目标**：${vm.goal}`);
    const exposure = formatProductExposure(plan);
    if (exposure) lines.push(`- **产品露出**：${exposure}`);
    lines.push('');

    vm.shots.forEach((shot) => {
      lines.push(`### ${label} · Shot ${shot.id}`);
      lines.push(`- **镜头描述**：${shot.desc}`);
      if (shot.sceneDescription) lines.push(`- **场景**：${shot.sceneDescription}`);
      if (shot.subjectDescription) lines.push(`- **角色 / 主体**：${shot.subjectDescription}`);
      if (shot.action) lines.push(`- **动作**：${shot.action}`);
      lines.push(`- **角色口播**：${shot.spokenText || '无角色口播'}`);
      lines.push(`- **旁白/画外音**：${shot.voiceoverText || '无旁白'}`);
      lines.push(`- **字幕文案**：${shot.subtitleText || '无字幕'}`);
      lines.push(`- **情绪**：${shot.emotion || '—'}`);
      lines.push(`- **时长**：${shot.duration}`);
      if (shot.cameraDescription) lines.push(`- **镜头运动**：${shot.cameraDescription}`);
      if (shot.imagePrompt) lines.push(`- **出图提示**：${shot.imagePrompt}`);
      if (shot.videoPrompt) lines.push(`- **视频提示**：${shot.videoPrompt}`);
      lines.push('');
    });
  });

  return lines.join('\n').trim() + '\n';
}

export function buildOverviewStoryboardMarkdown(pipeline: PipelineSummaryDto | null): string | null {
  if (!pipeline?.project) return null;
  const rows = pipeline.segment_scripts;
  if (!Array.isArray(rows) || rows.length === 0) return null;

  const name = safeProjectExportBaseName(pipeline.project.project_name);
  const bp = pipeline.story_blueprint?.blueprint ?? null;

  const lines: string[] = [];
  lines.push(`# ${name} — 分镜`);
  lines.push('');

  const sorted = [...rows].sort((a, b) => {
    const na = parseInt(String(a.segment_id).replace(/\D/g, ''), 10) || 0;
    const nb = parseInt(String(b.segment_id).replace(/\D/g, ''), 10) || 0;
    return na - nb;
  });

  sorted.forEach((row, idx) => {
    const vm = segmentScriptDtoToStepSegmentViewModel(row, idx);
    const plan = segmentPlanForIndex(bp, row.segment_id, idx);
    const label = `S${idx + 1}`;
    lines.push(`## Segment ${label} — ${vm.name}`);
    lines.push('');

    vm.shots.forEach((shot) => {
      lines.push(`### Shot ${shot.id}`);
      lines.push(`- **镜头描述**：${shot.desc}`);
      if (shot.sceneDescription) lines.push(`- **场景**：${shot.sceneDescription}`);
      if (shot.subjectDescription) lines.push(`- **角色**：${shot.subjectDescription}`);
      const exposure = formatProductExposure(plan);
      if (exposure) lines.push(`- **产品露出**：${exposure}`);
      lines.push(`- **时长**：${shot.duration}`);
      lines.push(`- **情绪**：${shot.emotion || '—'}`);
      lines.push('');
    });
  });

  return lines.join('\n').trim() + '\n';
}

File: backend/app/short_drama/services/overview_export_markdown.py
"""Markdown for Overview export (script / storyboard) — mirrors frontend field usage."""

from __future__ import annotations

import re
from typing import Any

from ..utils.segment_slots import normalize_segment_script_dict_for_read


def _safe_name(name: str) -> str:
    s = re.sub(r'[/\\?%*:|"<>]', "_", (name or "").strip())
    s = re.sub(r"\s+", " ", s).strip()
    return s or "project"


def _segment_plan_item(blueprint: dict | None, segment_id: str, index: int) -> dict | None:
    if not blueprint or not isinstance(blueprint, dict):
        return None
    plan = blueprint.get("segment_plan")
    if not isinstance(plan, list) or not plan:
        return None
    sid = (segment_id or "").strip()
    for p in plan:
        if isinstance(p, dict) and (p.get("segment_id") or "").strip() == sid:
            return p
    if 0 <= index < len(plan) and isinstance(plan[index], dict):
        return plan[index]
    return None


def _product_exposure(plan: dict | None) -> str | None:
    if not plan:
        return None
    m = (plan.get("product_exposure_mode") or "").strip()
    if m:
        return m
    s = (plan.get("summary") or "").strip()
    return s or None


def _fmt_duration_seconds(raw: Any) -> str:
    if raw is None or raw == "":
        return "—"
    try:
        f = float(raw)
        if f > 0 and f == int(f):
            return f"{int(f)}s"
        if f > 0:
            return f"{f}s"
    except (TypeError, ValueError):
        pass
    if isinstance(raw, str) and raw.strip():
        return raw.strip()
    return "—"


def _shot_desc(shot: dict, i: int) -> str:
    vd = (shot.get("visual_description") or "").strip()
    ad = (shot.get("action_description") or "").strip()
    if vd:
        return vd
    if ad:
        return ad
    return f"镜头 {i + 1}"


def build_script_markdown(
    *,
    project_name: str,
    project_row: Any,
    blueprint_json: dict | None,
    segment_rows: list[Any],
) -> str:
    name = _safe_name(project_name)
    lines: list[str] = [
        f"# {name} — 脚本",
        "",
        "## 项目信息",
        f"- **项目名**：{name}",
        f"- **格式**：{(getattr(project_row, 'format', None) or '—')}",
        f"- **风格**：{(getattr(project_row, 'style', None) or '—')}",
        f"- **视听风格**：{(getattr(project_row, 'visual_style', None) or '—')}",
        f"- **比例**：{(getattr(project_row, 'aspect_ratio', None) or '—')}",
        f"- **时长**：{(getattr(project_row, 'duration', None) or '—')}",
        "",
    ]

    bp = blueprint_json if isinstance(blueprint_json, dict) else None

    for idx, row in enumerate(segment_rows):
        raw = row.script_json if isinstance(getattr(row, "script_json", None), dict) else {}
        script = normalize_segment_script_dict_for_read(dict(raw))
        title = (script.get("title") or "").strip() or f"片段 {idx + 1}"
        goal = (script.get("goal") or "").strip() or "—"
        dl = script.get("duration_limit")
        dur = _fmt_duration_seconds(dl)
        plan = _segment_plan_item(bp, str(getattr(row, "segment_id", "") or ""), idx)
        exposure = _product_exposure(plan)

        label = f"S{idx + 1}"
        lines.append(f"## {label} — {title}")
        lines.append(f"- **标题**：{title}")
        lines.append(f"- **时长**：{dur}")
        lines.append(f"- **情绪 / 目标**：{goal}")
        if exposure:
            lines.append(f"- **产品露出**：{exposure}")
        lines.append("")

        shots = script.get("shots")
        if not isinstance(shots, list):
            shots = []
        if not shots:
            lines.append(f"### {label} · Shot 1")
            lines.append("- **镜头描述**：—")
            lines.append("")

        for si, raw in enumerate(shots):
            shot = raw if isinstance(raw, dict) else {}
            lines.append(f"### {label} · Shot {si + 1}")
            lines.append(f"- **镜头描述**：{_shot_desc(shot, si)}")
            sd = (shot.get("scene_description") or "").strip()
            if sd:
                lines.append(f"- **场景**：{sd}")
            subj = (shot.get("subject_description") or "").strip()
            if subj:
                lines.append(f"- **角色 / 主体**：{subj}")
            act = (shot.get("action_description") or "").strip()
            if act:
                lines.append(f"- **动作**：{act}")
            spoken = (shot.get("spoken_text") or shot.get("dialogue") or "").strip()
            voiceover = (shot.get("voiceover_text") or shot.get("voiceover") or shot.get("narration") or "").strip()
            subtitle = (shot.get("subtitle_text") or shot.get("subtitle") or "").strip()
            lines.append(f"- **角色口播**：{spoken or '无角色口播'}")
            lines.append(f"- **旁白/画外音**：{voiceover or '无旁白'}")
            lines.append(f"- **字幕文案**：{subtitle or '无字幕'}")
            em = (shot.get("emotion") or "").strip() or "—"
            lines.append(f"- **情绪**：{em}")
            lines.append(f"- **时长**：{_fmt_duration_seconds(shot.get('duration_seconds'))}")
            cd = (shot.get("camera_description") or "").strip()
            if cd:
                lines.append(f"- **镜头运动**：{cd}")
            ip = (shot.get("image_prompt") or "").strip()
            if ip:
                lines.append(f"- **出图提示**：{ip}")
            vp = (shot.get("video_prompt") or "").strip()
            if vp:
                lines.append(f"- **视频提示**：{vp}")
            lines.append("")

    return "\n".join(lines).rstrip() + "\n"


def build_storyboard_markdown(
    *,
    project_name: str,
    blueprint_json: dict | None,
    segment_rows: list[Any],
) -> str:
    name = _safe_name(project_name)
    lines: list[str] = [f"# {name} — 分镜", ""]
    bp = blueprint_json if isinstance(blueprint_json, dict) else None

    for idx, row in enumerate(segment_rows):
        raw = row.script_json if isinstance(getattr(row, "script_json", None), dict) else {}
        script = normalize_segment_script_dict_for_read(dict(raw))
        title = (script.get("title") or "").strip() or f"片段 {idx + 1}"
        plan = _segment_plan_item(bp, str(getattr(row, "segment_id", "") or ""), idx)
        exposure = _product_exposure(plan)
        label = f"S{idx + 1}"
        lines.append(f"## Segment {label} — {title}")
        lines.append("")

        shots = script.get("shots")
        if not isinstance(shots, list):
            shots = []
        if not shots:
            lines.append("### Shot 1")
            lines.append("- **镜头描述**：—")
            lines.append("")

        for si, raw in enumerate(shots):
            shot = raw if isinstance(raw, dict) else {}
            lines.append(f"### Shot {si + 1}")
            lines.append(f"- **镜头描述**：{_shot_desc(shot, si)}")
            sd = (shot.get("scene_description") or "").strip()
            if sd:
                lines.append(f"- **场景**：{sd}")
            subj = (shot.get("subject_description") or "").strip()
            if subj:
                lines.append(f"- **角色**：{subj}")
            if exposure:
                lines.append(f"- **产品露出**：{exposure}")
            lines.append(f"- **时长**：{_fmt_duration_seconds(shot.get('duration_seconds'))}")
            em = (shot.get("emotion") or "").strip() or "—"
            lines.append(f"- **情绪**：{em}")
            lines.append("")

    return "\n".join(lines).rstrip() + "\n"
